src/window.cpp

```diff
@@ -895,57 +874,100 @@ static bool MayBeShown(const Window *w)
+				/* Window 'v' and the dirty part of window 'w' do not intersect with each other */
+				continue;
 			}
 
-			if (top < (x = v->top)) {
-				DrawOverlappedWindow(w, left, top, right, x);
-				DrawOverlappedWindow(w, left, x, right, bottom);
-				return;
+			/*
+			 * (r >  vl && b >  vt && l <  vr && t <  vb)
+			 */
+			if (left < v_left) {
+				/* Window 'v' occludes window 'w'; split at left edge.
+				 * Draw the non-occluded part recursively, then fall through to trim the rest of the occluded parts.
+				 */
+				draw_overlapped(v->z_front, left, top, v_left, bottom);
+				left = v_left;
 			}
 
-			if (bottom > (x = v->top + v->height)) {
-				DrawOverlappedWindow(w, left, top, right, x);
-				DrawOverlappedWindow(w, left, x, right, bottom);
-				return;
+			/*
+			 * (r >  vl && b >  vt && l <  vr && t <  vb) && (l >= vl)
+			 * ==>
+			 * (vl <= l <  r && vt <  b && l <  vr && t <  vb)
+			 */
+			if (right > v_right) {
+				/* Window 'v' occludes window 'w'; split at right edge. */
+				draw_overlapped(v->z_front, v_right, top, right, bottom);
+				right = v_right;
 			}
 
+			/*
+			 * (vl <= l <  r && vt <  b && l <  vr && t <  vb) && (r <= vr)
+			 * ==>
+			 * (vl <= l <  r <= vr && vt <  b && t <  vb)
+			 */
+			if (top < v_top) {
+				/* Window 'v' occludes window 'w'; split at top edge. */
+				draw_overlapped(v->z_front, left, top, right, v_top);
+				top = v_top;
+			}
+
+			/*
+			 * (vl <= l <  r <= vr && vt <  b && t <  vb) && (t >= vt)
+			 * ==>
+			 * (vl <= l <  r <= vr && vt <= t <  b && t <  vb)
+			 */
+			if (bottom > v_bottom) {
+				/* Window 'v' occludes window 'w'; split at bottom edge. */
+				draw_overlapped(v->z_front, left, v_bottom, right, bottom);
+				bottom = v_bottom;
+			}
+
+			/*
+			 * (vl <= l <  r <= vr && vt <= t <  b && t <  vb) && (b <= vb)
+			 * ==>
+			 * (vl <= l <  r <= vr && vt <= t <  b <= vb)
+			 */
+
+			/* Window 'v' completely occludes (the dirty part of) window 'w' */
 			return;
 		}
-	}
 
-	/* Setup blitter, and dispatch a repaint event to window *wz */
-	DrawPixelInfo *dp = _cur_dpi;
-	dp->width = right - left;
-	dp->height = bottom - top;
-	dp->left = left - w->left;
-	dp->top = top - w->top;
-	dp->pitch = _screen.pitch;
-	dp->dst_ptr = BlitterFactory::GetCurrentBlitter()->MoveTo(_screen.dst_ptr, left, top);
-	dp->zoom = ZOOM_LVL_NORMAL;
-	w->OnPaint();
+		/* Setup blitter, and dispatch a repaint event to window *wz */
+		DrawPixelInfo *dp = _cur_dpi;
+		dp->width = right - left;
+		dp->height = bottom - top;
+		dp->left = left - w->left;
+		dp->top = top - w->top;
+		dp->pitch = _screen.pitch;
+		dp->dst_ptr = BlitterFactory::GetCurrentBlitter()->MoveTo(_screen.dst_ptr, left, top);
+		dp->zoom = ZOOM_LVL_NORMAL;
+		w->OnPaint();
+		if (gfx_dirty) {
+			VideoDriver::GetInstance()->MakeDirty(left, top, right - left, bottom - top);
+			UnsetDirtyBlocks(left, top, right, bottom);
+		}
+	};
+	draw_overlapped(w->z_front, left, top, right, bottom);
```

Comment:
Does C++ support self-invoking lambdas? If so, this line may be unnecessary. I'd certainly prefer to have it near the top of the function, if that's at all possible.